interpreter/statement.go

```diff
@@ -494,3 +503,94 @@ func (i *Interpreter) ProcessIfStatement(stmt *ast.IfStatement, ds DebugState) (
 	}
 	return NONE, nil
 }
+
+func (i *Interpreter) ProcessSwitchStatement(stmt *ast.SwitchStatement, ds DebugState) (State, error) {
+	// User defined functions used in a switch control statement must have a STRING return type.
+	if fnCall, ok := stmt.Control.(*ast.FunctionCallExpression); ok {
+		if fn, ok := i.ctx.SubroutineFunctions[fnCall.Function.Value]; ok {
+			if fn.ReturnType.Value != string(value.StringType) {
+				return NONE, errors.WithStack(exception.Runtime(
+					&fnCall.Token,
+					"user defined function has invalid return type (%s) for switch, must be STRING",
+					fn.ReturnType.Value,
+				))
+			}
+		}
+	}
```

Comment:
We don't want to return early in this section if the function isn't found as this block is for validating the return type only if the function is user defined. If it doesn't exist in `SubroutineFunctions` it could just mean the function is a builtin. The call to `i.ProcessExpression` after this section will handle the case of the function not being defined.

That said this could still have one layer of nesting removed.
```
fn, ok := i.ctx.SubroutineFunctions[fnCall.Function.Value]
if ok && fn.ReturnType.Value != string(value.StringType) {
   // ...
}
```

Reply:
Do you mean the `ast.FunctionCallStatement` is a possibility to be a built-in function? if so, you are right